File: Algorithms/DepthFirstSearch/Graph.py
from typing import *

class Edge:
  def __init__(self, fr:int, to:int, wt:int = 1) -> None:
    self.fr:int = fr
    self.to:int = to
    self.wt:int = wt

class Graph:
  def __init__(self, edgelist:List[Edge]|None = None) -> None:
    self.adjacencyList:Dict[int,Set[Edge]] = dict()
    if edgelist == None: return
    for edge in edgelist:
      self.AddNode(edge.fr)
      self.AddNode(edge.to)
      self.AddEdge(edge)
    return

  def AddNode(self, node:int) -> None:
    if node in self.adjacencyList: return
    self.adjacencyList[node] = set()
    return

  def AddEdge(self, edge) -> None:
    self.adjacencyList[edge.fr].add(edge)
    return

  def PrintAdjacencyList(self) -> None:
    for element in self.adjacencyList:
      print(element, sep="", end=" -> ")
      for edge in self.adjacencyList[element]:
        print(f"{edge.to}({edge.wt})", sep="", end=" -> ")
      print("X")
    return

  def DFS(self, root:int|None = None) -> List[int]:
    sequence:List[int] = []
    def traverse(root: int) -> None:
      if root in sequence: return
      sequence.append(root)
      for edge in self.adjacencyList[root]:
        traverse(edge.to)
      return
    traverse(root or list(self.adjacencyList.keys())[0])
    return sequence

edgelist = [
  Edge(1, 2),
  Edge(1, 3), 
  Edge(1, 4),
  Edge(2, 3),
  Edge(3, 4),
  Edge(3, 5),
  Edge(4, 5),
]

gr = Graph(edgelist)

gr.PrintAdjacencyList()

print(gr.DFS())

File: Algorithms/DepthFirstSearch/Graph.cpp
#include <iostream>
#include <map>
#include <vector>
#include <set>
#include <unordered_map>
#include <unordered_set>

struct Edge
{
  int fr;
  int to;
  int wt;

  Edge(int fr, int to, int wt = 1)
    : fr(fr), to(to), wt(wt) { }

  Edge Copy()
  {
    return Edge(fr, to, wt);
  }
};

struct Graph
{
  std::unordered_map<int, std::vector<Edge>> adjacenclyList;

  Graph() { }

  Graph(std::vector<Edge> edgelist)
  {
    for (Edge& edge : edgelist)
    {
      AddEdge(edge);
    }
  }

  void AddNode(int node)
  {
    if (adjacenclyList.find(node) != adjacenclyList.end()) return;
    adjacenclyList[node] = std::vector<Edge>();
  }

  void AddEdge(Edge& edge)
  {
    AddNode(edge.fr);
    AddNode(edge.to);
    adjacenclyList[edge.fr].emplace_back(edge.Copy());
  }

  void PrintAdjacencyList()
  {
    for (std::pair<int, std::vector<Edge>> pair : adjacenclyList)
    {
      std::cout << pair.first << " : ";
      for (Edge edge : pair.second)
      {
        std::cout << edge.to << "(" << edge.wt << ")" << " -> ";
      }
      std::cout << "X\n";
    }
  }

  std::vector<int> GetDFSSequence()
  {
    int first;
    for (auto pair : adjacenclyList) 
    {
      first = pair.first;
      break;
    }
    return GetDFSSequence(first);
  }

  std::vector<int> GetDFSSequence(int root)
  {
    std::vector<int> sequence;
    dfsTraversal(sequence, root);
    return sequence;
  }

private:
  void dfsTraversal(std::vector<int>& sequence, int root)
  {
    if (std::find(sequence.begin(), sequence.end(), root) != sequence.end()) return;
    sequence.push_back(root);
    for (Edge edge : adjacenclyList[root]) {
      dfsTraversal(sequence, edge.to);
    }
  }
};

int main()
{
  std::vector<Edge> edgelist = {
    Edge(1, 2),
    Edge(1, 3),
    Edge(1, 4),
    Edge(2, 3),
    Edge(3, 4),
    Edge(3, 5),
    Edge(4, 5),
  };

  Graph gr(edgelist);

  gr.PrintAdjacencyList();

  for (int val : gr.GetDFSSequence())
  {
    std::cout << val << " ";
  }
  std::cout << "\n";

  return 0;
}